client/etcd/client.go

```diff
@@ -199,27 +220,33 @@ func newClient(endpoints []string) (*clientv3.Client, error) {
 	return clientv3.New(clientv3.Config{Endpoints: endpoints})
 }
 
-func cacheFileForZone(cacheDir, service, zone string) string {
-	if cacheDir == "" || service == "" || zone == "" {
-		return ""
-	}
-	return filepath.Join(cacheDir, fileName(service, zone))
-}
-
-func fileName(service, zone string) string {
-	cacheFileName := fmt.Sprintf(cacheFileFormat, service, zone)
-
-	return strings.Replace(cacheFileName, string(os.PathSeparator), "_", -1)
+func (c *csclient) cacheFileFn(zone string) etcdkv.CacheFileFn {
+	return etcdkv.CacheFileFn(func(namespace string) string {
+		if c.opts.CacheDir() == "" || c.opts.Service() == "" {
+			return ""
+		}
+		return filepath.Join(c.opts.CacheDir(), fileName(namespace, c.opts.Service(), zone))
+	})
 }
 
-func prefix(env string) string {
-	res := kvPrefix
-	if env != "" {
-		res = concat(res, concat(env, keySeparator))
+func fileName(parts ...string) string {
+	// get non-empty parts
+	idx := 0
+	for _, part := range parts {
+		if part == "" {
+			continue
+		}
+		parts[idx] = part
```

Comment:
nit: could compare the part index with `idx` here and don't set it if its' the same